### PR TITLE
[20.09, staging] glibc: fix CVE-2020-6096

### PR DESCRIPTION
**Pull request by redvers**
###### Motivation for this change

Addresses: #93992 - CVE-2020-6092

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

I do not own an arm7 system to test the build on - but my local x86_64 system continues to be happy.

### CODE REVIEW

File: pkgs/development/libraries/glibc/common.nix [hunk 1]
```diff
@@ -115,6 +115,8 @@ stdenv.mkDerivation ({
       ./fix-x64-abi.patch
       ./2.30-cve-2020-1752.patch
       ./2.31-cve-2020-10029.patch
+      ./2.31-cve-2020-6096.0.patch
+      ./2.31-cve-2020-6096.1.patch
```
Comment (mweinelt) [inline pkgs/development/libraries/glibc/common.nix:119]: Any particular reason why we don't use fetchurl to grab those patches?

**Review by mweinelt — CHANGES_REQUESTED** [review]
The commit message needs some work, I'm sorry if I was not clear enough on what to put in.

I think something like this is more helpful:

```
glibc: fix CVE-2020-6096

Fixes a signed comparison vulnerability in the ARMv7 memcpy and memmove functions.

https://sourceware.org/bugzilla/show_bug.cgi?id=25620

Fixes: CVE-2020-6096
```

Attributes like `Fixes`, `Reviewed-by`, `Tested-by` et al. are usually put at the end of the commit message.

File: pkgs/development/libraries/glibc/common.nix (hunk 1 above)
Comment (andir) [reply pkgs/development/libraries/glibc/common.nix:119]: bootstrap can't use fetchpatch and fetchurl might not be stable.

Comment (redvers) [reply pkgs/development/libraries/glibc/common.nix:119]: Yes, instructions in common.nix states to download them directly as they're a part of the bootstrap process:

```
# Note: this package is used for bootstrapping fetchurl, and thus
# cannot use fetchpatch! All mutable patches (generated by GitHub or
# cgit) that are needed here should be included directly in Nixpkgs as
# files.
```
https://github.com/NixOS/nixpkgs/blob/31487784f9ae66e715b2df1e7727119890d05a39/pkgs/development/libraries/glibc/common.nix#L20-L23

**Comment by redvers** [comment]
Got it - I'll get that commit message massaged - one moment caller...

**Comment by redvers** [comment]
@mweinelt  - when you say "Reviewed-by:" and "Tested-by:" there's no entry for that in the commits for these two patches
(Although there is for the associated patch for the removal of the test).

I've included direct links to the patches.

**Comment by mweinelt** [comment]
This also affects memmove. Check https://sourceware.org/bugzilla/show_bug.cgi?id=25620#c27.

**Comment by redvers** [comment]
Yup - that patch is included - but I'll update the commit message to include it.

**Comment by mweinelt** [comment]
Is this already mitigated in master? If yes, how and where? Why can't we backport that solution?

**Comment by andir** [comment]
For what it's worth this looks fine. I've build the small release set of NixOS against this PR and only the scipy build fails (which is IIRC unrelated to this change): https://hydra.h4ck.space/build/338254#tabs-constituents

**Comment by redvers** [comment]
> Is this already mitigated in master? If yes, how and where? Why can't we backport that solution?

Yes, it was mitigated by bumping to 2.32 which I did open as a cherry-picked PR upon which I received my first ever thumbs-down on a PR :-)

Apparently the bump to 2.32 has so many changes there's still stuff in master that is broken.

**Comment by vcunat** [comment]
Oh, I see you [decided](https://github.com/NixOS/nixpkgs/pull/103539) for a single patch instead of the whole upstream 2.31 branch?  I hope we won't be missing some important fix, but anyway... such changes can come later.

Detail: our aarch64 rebuild will take quite a long time, considering that the platform isn't said to be affected, but if you don't hurry, the non-conditional patching does appear cleanr.

**Comment by redvers** [comment]
@vcunat  - after discussions on #nixos-security we decided to pull the minimal changeset to address the security vulnerability.

There's so much domain knowledge required to make a judgement call as to which patches are fixes and which are 'fixes that will break things' that we took the safest path.

The provided patches only patch the arm arch specific files - the safest option.  As you said, other potential patches are better left for later in the hands of someone better equipped to deal with that domain.

**Comment by vcunat** [comment]
Sure, I never suggested going through all the patches.  Take the whole branch as upstream picked which patches were considered suitable for the stable branch &ndash; or just take some specific ones like you decided to do.

**Pull request merged**